fluentinterface/src/main/java/com/iluwatar/fluentinterface/fluentiterable/lazy/LazyFluentIterable.java

```diff
@@ -0,0 +1,211 @@
+                            return candidate;
+                        }
+
+                        return null;
+                    }
+                };
+            }
+        };
+    }
+
+    /**
+     * Can be used to collect objects from the iteration. Is a terminating operation.
+     * @return an Optional containing the first object of this Iterable
+     */
+    @Override
+    public Optional<TYPE> first() {
+        Iterator<TYPE> resultIterator = first(1).iterator();
+        return resultIterator.hasNext() ? Optional.of(resultIterator.next()) : Optional.empty();
+    }
+
+    /**
+     * Can be used to collect objects from the iteration.
+     * @param count defines the number of objects to return
+     * @return the same FluentIterable with a collection decimated to a maximum of 'count' first objects.
+     */
+    @Override
+    public FluentIterable<TYPE> first(int count) {
+        return new LazyFluentIterable<TYPE>() {
+            @Override
+            public Iterator<TYPE> iterator() {
+                return new DecoratingIterator<TYPE>(iterable.iterator()) {
+                    int currentIndex = 0;
+
+                    @Override
+                    public TYPE computeNext() {
+                        if(currentIndex < count) {
+                            if(fromIterator.hasNext()) {
+                                TYPE candidate = fromIterator.next();
+                                currentIndex++;
+                                return candidate;
+                            }
+                        }
+                        return null;
+                    }
+                };
+            }
+        };
+    }
+
+    /**
+     * Can be used to collect objects from the iteration. Is a terminating operation.
+     * @return an Optional containing the last object of this Iterable
+     */
+    @Override
+    public Optional<TYPE> last() {
+        Iterator<TYPE> resultIterator = last(1).iterator();
+        return resultIterator.hasNext() ? Optional.of(resultIterator.next()) : Optional.empty();
+    }
+
+    /**
+     * Can be used to collect objects from the Iterable. Is a terminating operation.
+     * This operation is memory intensive, because the contents of this Iterable
+     * are collected into a List, when the next object is requested.
+     * @param count defines the number of objects to return
+     * @return the same FluentIterable with a collection decimated to a maximum of 'count' last objects
+     */
+    @Override
+    public FluentIterable<TYPE> last(int count) {return new LazyFluentIterable<TYPE>() {
+        @Override
+        public Iterator<TYPE> iterator() {
+            return new DecoratingIterator<TYPE>(iterable.iterator()) {
+                int currentIndex = 0;
+
+                @Override
+                public TYPE computeNext() {
+                    List<TYPE> list = new ArrayList<>();
+
+                    Iterator<TYPE> newIterator = iterable.iterator();
```

Comment:
We don't need to populate the list on each `computeNext()`. You can cache it on first call and then utilize it. If you have doubts then I am implementing it, will give you sample.